spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/RequestRateLimiterGatewayFilterFactory.java

```diff
@@ -96,42 +95,39 @@ public GatewayFilter apply(Config config) {
 		HttpStatusHolder emptyKeyStatus = HttpStatusHolder
 				.parse(getOrDefault(config.emptyKeyStatus, this.emptyKeyStatusCode));
 
-		return (exchange, chain) -> {
-			Route route = exchange
-					.getAttribute(ServerWebExchangeUtils.GATEWAY_ROUTE_ATTR);
-
-			return resolver.resolve(exchange).defaultIfEmpty(EMPTY_KEY).flatMap(key -> {
-				if (EMPTY_KEY.equals(key)) {
-					if (denyEmpty) {
-						setResponseStatus(exchange, emptyKeyStatus);
-						return exchange.getResponse().setComplete();
-					}
-					return chain.filter(exchange);
-				}
-				return limiter.isAllowed(route.getId(), key).flatMap(response -> {
-
-					for (Map.Entry<String, String> header : response.getHeaders()
-							.entrySet()) {
-						exchange.getResponse().getHeaders().add(header.getKey(),
-								header.getValue());
-					}
-
-					if (response.isAllowed()) {
+		return (exchange, chain) -> resolver.resolve(exchange).defaultIfEmpty(EMPTY_KEY)
+				.flatMap(key -> {
+					if (EMPTY_KEY.equals(key)) {
+						if (denyEmpty) {
+							setResponseStatus(exchange, emptyKeyStatus);
+							return exchange.getResponse().setComplete();
+						}
 						return chain.filter(exchange);
 					}
-
-					setResponseStatus(exchange, config.getStatusCode());
-					return exchange.getResponse().setComplete();
+					return limiter.isAllowed(config.getRouteId(), key)
```

Comment:
If `config.getRouteId() == null` use `route.getId()`.

Reply:
I'm not sure if that would ever be possible from the gateway's perspective since the config implements `HasRouteId`, therefor the config will always get an id from the `RouteDefinitionRouteLocator`. Is there another path in the code where the `apply()` function is being called?

Reply:
Oh, since it's set there.